Update/deployment guides v2 (#4068)
circleci/circleci-docs: jekyll/_cci2/deployment-integrations.adoc

* **Deployment Parameters**: You will need to add environment variables/SSH keys to fulfill your deployment steps and provide autorisation details and parameters.
+
These can be added to the project itself (follow links below for detailed instructions), or defined within your configuration. For an example of defining environment variables within your configuration see the non-orb Heroku example below – within the deploy job configuration we have:
+
```
environment:
      HEROKU_APP: "sleepy-refuge-55486"
```
** Add project-level <<env-vars#setting-an-environment-variable-in-a-project,environment variables>>
** Add project-level <<add-ssh-key#,SSH keys>> 

* **Manual Approval**: If your deployment strategy requires a manual approval step, you can include a <<configuration-reference#type,hold/approve>> job within your workflow. A manual approval button will then be available from the workflows map in the CircleCI UI.

== Using Orbs to Simplify Deployment

> https://circleci.com/docs/2.0/orb-intro/[Orbs] are available for many common deployment targets to help simplify and streamline your config – check out the full range of available orbs in the https://circleci.com/orbs/registry/[Orbs Registry]. Orbs are packages of reusable configuration, and often, for simple pipelines, simply invoking an orb, including a pre-configured job, and setting any required environment variables/SSH keys within your project will get you the results you need with no further configuration. 

Author: ```suggestion
https://circleci.com/docs/2.0/orb-intro/[Orbs] are available for many common deployment targets to help simplify and streamline your config – check out the full range of available orbs in the https://circleci.com/orbs/registry/[Orbs Registry]. Orbs are packages of reusable configuration. For simple pipelines, simply invoking an orb, including a pre-configured job, and setting any required environment variables/SSH keys within your project will get you the results you need with no further configuration. 
```

//// 
.Deployment
image::pipeline-to-deployment.png[Deployment]
////

== Deployment: The Basics

* **Deployment Job**: To deploy your application, add a <<job-steps#jobs-overview,job>> to your `.circleci/config.yml` file and configure the job to run the steps you require. The <<deployment-examples#,deployment examples>> document covers example config for many common deployment targets.

* **Deployment Parameters**: You will need to add environment variables/SSH keys to fulfill your deployment steps and provide autorisation details and parameters.
+
These can be added to the project itself (follow links below for detailed instructions), or defined within your configuration. For an example of defining environment variables within your configuration see the non-orb Heroku example below – within the deploy job configuration we have:
+
```
environment:
      HEROKU_APP: "sleepy-refuge-55486"
```
** Add project-level <<env-vars#setting-an-environment-variable-in-a-project,environment variables>>
** Add project-level <<add-ssh-key#,SSH keys>> 

> * **Manual Approval**: If your deployment strategy requires a manual approval step, you can include a <<configuration-reference#type,hold/approve>> job within your workflow. A manual approval button will then be available from the workflows map in the CircleCI UI.

Author: Add in a bullet about using workflows (api?) here, eventually with a link to Michelle's optimization section about workflow/deployment optimization.

== Using Orbs to Simplify Deployment

> https://circleci.com/docs/2.0/orb-intro/[Orbs] are available for many common deployment targets to help simplify and streamline your config – check out the full range of available orbs in the https://circleci.com/orbs/registry/[Orbs Registry]. Orbs are packages of reusable configuration, and often, for simple pipelines, simply invoking an orb, including a pre-configured job, and setting any required environment variables/SSH keys within your project will get you the results you need with no further configuration. 

Reviewer: "Orbs are packages of reusable configuration, and often, for simple pipelines, simply invoking an orb, including a pre-configured job, and setting any required environment variables/SSH keys within your project will get you the results you need with no further configuration. "

opinion: I would suggest breaking this into two sentences to reduce the use of commas.

Author: ha yeah I kept reading that and thinking I need to fix it. So many commas!

==== 2.0 config

The configuration uses https://circleci.com/docs/2.0/workflows/[workflows] to deploy only if the `sequential-branch-filter` branch is checked out and the `build` job has run. If your deploy job uses any output from previous jobs, you can share that data by https://circleci.com/docs/2.0/workflows/#using-workspaces-to-share-data-among-jobs[using workspaces]. For more information on conditional deploys, see https://circleci.com/docs/2.0/workflows/#using-contexts-and-filtering-in-your-workflows[Using Contexts and Filtering in your Workflows].

```yaml
version: 2.0

jobs:
  build:
    docker:
      - image: circleci/ruby:2.4-node # primary container - where job steps are run
      - image: circleci/postgres:9.4.12-alpine # services container
    working_directory: ~/circleci-demo-workflows
    steps:
      - checkout
      # Bundle install dependencies
      - run: bundle install --path vendor/bundle

      # Database setup
      - run: bundle exec rake db:create db:schema:load

      - run:
          name: Run tests
          command: rake

  deploy:
    machine:
        enabled: true
    working_directory: ~/circleci-demo-workflows
    environment:

>       HEROKU_APP: "sleepy-refuge-55486" # define env vae $HEROKU_APP

Reviewer: ```suggestion
      HEROKU_APP: "sleepy-refuge-55486" # define env var $HEROKU_APP
```

toc::[]

[discrete]
== Overview

Once a software application has been developed and tested, it needs to be deployed and made available for its intended audience. CircleCI can deploy to virtualy any target, and can be easily configured to integrate with other services for QA/testing, feature management, and deployment strategies such as blue-green or canary deployment. Quickly and easily customize your config to match your requirements, whether a fully automated process or elements of manual approval are required.
//// 
.Deployment
image::pipeline-to-deployment.png[Deployment]
////

== Deployment: The Basics

* **Deployment Job**: To deploy your application, add a <<jobs-steps#jobs-overview,job>> to your `.circleci/config.yml` file and configure the job to run the steps you require. The <<deployment-examples#,deployment examples>> document covers example config for many common deployment targets.

* **Deployment Parameters**: You may need to add environment variables and SSH keys to fulfill your deployment steps.
+

> Environment variables can be added to the project itself via the CircleCI UI (follow links below for detailed instructions), or defined within your configuration. An example of defining environment variables _within_ your configuration can be seen <<2.0-config,further down this page within the deploy job config for the non-orb Heroku example>>.

Author: ```suggestion
Environment variables can be added to the project itself via the CircleCI UI (follow links below for detailed instructions), or defined within your configuration. An example of defining environment variables _within_ your configuration can be seen further down this page within the deploy job config for the non-orb Heroku example.
```

> Environment variables can be added to the project itself via the CircleCI UI (follow links below for detailed instructions), or defined within your configuration. An example of defining environment variables _within_ your configuration can be seen <<#2.0-config,further down this page within the deploy job config for the non-orb Heroku example>>.

Author: ```suggestion
Environment variables can be added to the project itself via the CircleCI UI (follow links below for detailed instructions), or defined within your configuration. An example of defining environment variables _within_ your configuration can be seen <<2.0-config,further down this page within the deploy job config for the non-orb Heroku example>>.
```

:toc: macro
:toc-title:

Deployment at CircleCI is just the same as any other job configuration; you create a job and configure it to deploy to virtually any target. This document provides an overview of the deployment process, along with best practices and optimization strategies.

toc::[]

[discrete]
== Overview

Once a software application has been developed and tested, it needs to be deployed and made available for its intended audience. CircleCI can deploy to virtualy any target, and can be easily configured to integrate with other services for QA/testing, feature management, and deployment strategies such as blue-green or canary deployment. Quickly and easily customize your config to match your requirements, whether a fully automated process or elements of manual approval are required.
//// 
.Deployment
image::pipeline-to-deployment.png[Deployment]
////

== Deployment: The Basics

* **Deployment Job**: To deploy your application, add a <<jobs-steps#jobs-overview,job>> to your `.circleci/config.yml` file and configure the job to run the steps you require. The <<deployment-examples#,deployment examples>> document covers example config for many common deployment targets.

> * **Deployment Parameters**: You will need to add environment variables/SSH keys to fulfill your deployment steps and provide authorization details and parameters.

Reviewer 2: I had trouble parsing this sentence. I think it'd be clearer in the form, "You may need to add something, something, or something".

Author: Thanks, fixed.

> * **Deployment Parameters**: You will need to add environment variables/SSH keys to fulfill your deployment steps and provide autorisation details and parameters.

Author: Good spot! Thanks

> * **Deployment Parameters**: You will need to add environment variables/SSH keys to fulfill your deployment steps and provide autorisation details and parameters.

Author: ```suggestion
* **Deployment Parameters**: You will need to add environment variables/SSH keys to fulfill your deployment steps and provide authorization details and parameters.
```

+

> Environment variables can be added to the project itself (follow links below for detailed instructions), or defined within your configuration. An example of defining environment variables _within_ your configuration can be seen below within the deploy job config for the non-orb Heroku example:

Reviewer 2: Does this come from a larger example? My instinct reading this would be to put `environment` at the root level of my config file, but I'm pretty sure it is actually supposed to be nested. Maybe this could be a link to a section with a larger example, like the ones below?

Author: Good point, that is confusing - fixed!

+
```
environment:
      HEROKU_APP: "sleepy-refuge-55486"
```
** Add project-level <<env-vars#setting-an-environment-variable-in-a-project,environment variables>>
** Add project-level <<add-ssh-key#,SSH keys>> 

* **Manual Approval**: If your deployment strategy requires a manual approval step, you can include a <<workflows#holding-a-workflow-for-a-manual-approval,hold/approve>> job within your workflow. A manual approval button will then be available from the workflows map in the CircleCI UI.

== Using Orbs to Simplify Deployment

> At CircleCI we offer packages of reusable configuration, known as <<orb-intro#,orbs>>. Orbs are available for many common deployment targets, which can help you to simplify and streamline your config. Check out the full range of available orbs in the https://circleci.com/orbs/registry/[Orbs Registry]. For simple pipelines simply invoking an orb, including a pre-configured job, and setting any required environment variables/SSH keys within your project, will get you the results you need with no further configuration. 

Reviewer 2: The last sentence here is hard to follow. I think the sentiment you're trying to communicate is, "For simple deployment processes, an orb will get you the results you need with minimal configuration".

Author: Yes I agree there's no need for that sentence to be so complicated. Fixed.

> As an example, consider the https://circleci.com/orbs/registry/orb/circleci/aws-code-deploy[AWS CodeDeploy orb]. This orb has a pre-configured `deploy` job, which you can include in your configuration with a single line `aws-code-deploy/deploy` once the orb has been invoked by adding the orbs stanza. For this example deployment can be as simple as:

Reviewer 2: This is a great example.

```yaml
version: 2.1

orbs:
  aws-code-deploy: circleci/aws-code-deploy@0.0.11

workflows:
  deploy_application:
    jobs:
      - aws-code-deploy/deploy:
          application-name: myApplication
          deployment-group: myDeploymentGroup
          service-role-arn: myDeploymentGroupRoleARN
          bundle-bucket: myApplicationS3Bucket
          bundle-key: myS3BucketKey
```

Under the hood, this orb creates, bundles and deploys your application using your specified parameters set under the `aws-code-deploy/deploy` job declaration.

== A Simple Example Using Heroku

> Below is a simple example of deploying a Rails application to Heroku. 

Author: ```suggestion
In this section we will look at a simple example of deploying a Rails application to Heroku. 
```
